File: backend/app/models/product.py
"""
Modelos de Producto y Categoría.
"""

from datetime import datetime, timezone

from sqlalchemy import (
    Column,
    Integer,
    String,
    Boolean,
    DateTime,
    Float,
    ForeignKey,
    Text,
)
from sqlalchemy.orm import relationship

from app.database import Base


class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False, unique=True)
    description = Column(String(255), nullable=True)
    icon = Column(String(10), default="🍗")  # Emoji icon
    is_active = Column(Boolean, default=True)
    display_order = Column(Integer, default=0)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    # Relationships
    products = relationship("Product", back_populates="category")

    def __repr__(self):
        return f"<Category {self.name}>"


class Product(Base):
    __tablename__ = "products"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    price = Column(Float, nullable=False)
    category_id = Column(Integer, ForeignKey("categories.id"), nullable=False)
    icon = Column(String(10), default="🍗")  # Emoji icon
    image_url = Column(String(500), nullable=True)
    is_available = Column(Boolean, default=True)
    is_featured = Column(Boolean, default=False)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    # Relationships
    category = relationship("Category", back_populates="products")
    order_items = relationship("OrderItem", back_populates="product")

    def __repr__(self):
        return f"<Product {self.name} S/{self.price}>"

File: mobile/lib/models/product.dart
class Category {
  Category({
    required this.id,
    required this.name,
    this.description,
    required this.icon,
    required this.isActive,
    required this.displayOrder,
  });

  final int id;
  final String name;
  final String? description;
  final String icon;
  final bool isActive;
  final int displayOrder;

  factory Category.fromJson(Map<String, dynamic> j) => Category(
        id: j['id'] as int,
        name: j['name'] as String? ?? '',
        description: j['description'] as String?,
        icon: j['icon'] as String? ?? '🍽️',
        isActive: j['is_active'] as bool? ?? true,
        displayOrder: (j['display_order'] as num?)?.toInt() ?? 0,
      );
}

class Product {
  Product({
    required this.id,
    required this.name,
    this.description,
    required this.price,
    required this.categoryId,
    required this.icon,
    this.imageUrl,
    required this.isFeatured,
    required this.isAvailable,
    this.createdAt,
    this.category,
  });

  final int id;
  final String name;
  final String? description;
  final double price;
  final int categoryId;
  final String icon;
  final String? imageUrl;
  final bool isFeatured;
  final bool isAvailable;
  final DateTime? createdAt;
  final Category? category;

  factory Product.fromJson(Map<String, dynamic> j) => Product(
        id: j['id'] as int,
        name: j['name'] as String? ?? '',
        description: j['description'] as String?,
        price: (j['price'] as num?)?.toDouble() ?? 0,
        categoryId: (j['category_id'] as num?)?.toInt() ?? 0,
        icon: j['icon'] as String? ?? '🍗',
        imageUrl: j['image_url'] as String?,
        isFeatured: j['is_featured'] as bool? ?? false,
        isAvailable: j['is_available'] as bool? ?? true,
        createdAt: DateTime.tryParse(j['created_at'] as String? ?? ''),
        category: j['category'] is Map<String, dynamic>
            ? Category.fromJson(j['category'] as Map<String, dynamic>)
            : null,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'price': price,
        'category_id': categoryId,
        'icon': icon,
        'image_url': imageUrl,
        'is_featured': isFeatured,
        'is_available': isAvailable,
      };
}
